ledger/ledger-api-common/src/main/scala/com/digitalasset/platform/server/api/authorization/auth/AuthServiceJWT.scala

```diff
@@ -0,0 +1,97 @@
+// Copyright (c) 2019 The DAML Authors. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package com.digitalasset.platform.server.api.authorization.auth
+
+import java.util.concurrent.{CompletableFuture, CompletionStage}
+
+import com.daml.ledger.participant.state.v1.{AuthService, Claim, ClaimActAsParty, ClaimAdmin, ClaimPublic, Claims}
+import com.digitalasset.daml.lf.data.Ref
+import io.grpc.Metadata
+import org.slf4j.{Logger, LoggerFactory}
+import spray.json._
+
+import scala.collection.mutable.ListBuffer
+import scala.util.Try
+
+/** The JWT token payload used in [[AuthServiceJWT]]
+  *
+  * TODO(RA): Do we need to wrap everything in a "com.daml.ledger.api" field,
+  * to prevent name collisions with other JWT fields?
+  *
+  * For forward/backward compatibility reasons, all fields are optional.
+  */
+case class AuthServiceJWTPayload(
+  ledgerId: Option[String],
+  applicationId: Option[String],
+  admin: Option[Boolean],
+  actAsParty: Option[List[String]]
+)
+
+object AuthServiceJWTProtocol extends DefaultJsonProtocol {
+  implicit val jwtPayloadFormat = jsonFormat4(AuthServiceJWTPayload.apply)
+}
+
+/** An AuthService that reads a JWT token from a `Authorization: Bearer` HTTP header.
+  * The token is expected to use the format as defined in [[AuthServiceJWTPayload]]:
+  */
+class AuthServiceJWT() extends AuthService {
+
+  protected val logger: Logger = LoggerFactory.getLogger(AuthServiceJWT.getClass)
+
+  override def decodeMetadata(headers: Metadata): CompletionStage[Claims] = {
+    val tokenE = for {
+      headerValue <- Option.apply(headers.get(AuthServiceJWT.AUTHORIZATION_KEY))
+        .toRight(new RuntimeException("Authorization header not found"))
+      token <- AuthServiceJWT.decodeAndParse(headerValue)
+    } yield token
+
+    tokenE.fold(
+      error => {
+        logger.warn("Authorization error:", error)
+        CompletableFuture.completedFuture(Claims.empty)
+      },
+      token => CompletableFuture.completedFuture(AuthServiceJWT.payloadToClaims(token))
+    )
+  }
+}
+
+object AuthServiceJWT {
+  val AUTHORIZATION_KEY: Metadata.Key[String] =
+    Metadata.Key.of("Authorization", Metadata.ASCII_STRING_MARSHALLER)
+
+  def apply() = new AuthServiceJWT()
+
+  def parsePayload(jwtPayload: String): Either[Throwable, AuthServiceJWTPayload] = {
+    import AuthServiceJWTProtocol._
+    Try(JsonParser(jwtPayload).convertTo[AuthServiceJWTPayload]).toEither
+  }
+
+  def decodeAndParse(headerValue: String): Either[Throwable, AuthServiceJWTPayload] = {
+    val bearerTokenRegex = "[Bb]earer: (.*)".r
+    for {
+      token <- bearerTokenRegex.findFirstIn(headerValue)
+        .toRight(new RuntimeException("Authorization header does not use Bearer format"))
+      decoded <- com.digitalasset.jwt.JwtDecoder.decode(com.digitalasset.jwt.domain.Jwt(token))
```

Comment:
you need to use `com.digitalasset.jwt.JwtVerifier` here. Decoder does not do any validation, only decodes the payload without checking the JWT signature. To create `JwtVerifier` with secret, use `com.digitalasset.jwt.HMAC256Verifier`, see example (removed from master):
https://github.com/digital-asset/daml/blob/c7b5482e43a0810bde96a7ecb6779d5ad973a75d/ledger-service/http-json/src/main/scala/com/digitalasset/http/HttpService.scala#L84-L86

Reply:
and yeah you probably don't want to hardcode "secret" in here like I did :)

Reply:
Thanks for the link. Originally, we wanted to split the work into two steps: The first version of the sandbox JWT authorization was to come without signature verification. In a next step, we would add JWT signature verification and the corresponding CLI parameters to set up the secret (HS256) or public key (RS256).